iceoryx_utils/include/iceoryx_utils/internal/lockfree_queue/index_queue.inl

```diff
@@ -0,0 +1,274 @@
+
+        if (cellIsValidToRead)
+        {
+            // case (1)
+            if (tryToGainOwnershipAt(readPosition))
+            {
+                break; // pop successful
+            }
+            // retry, readPosition was already updated via CAS in tryToGainOwnershipAt
+        }
+        else
+        {
+            // readPosition is ahead by one cycle, queue was empty at loadValueAt(...)
+            auto isEmpty = value.isOneCycleBehind(readPosition);
+
+            if (isEmpty)
+            {
+                // case (2)
+                return false;
+            }
+
+            // case (3) and (4) requires loading readPosition again
+            readPosition = loadNextReadPosition();
+        }
+
+        // readPosition is outdated, retry operation
+
+    } while (!ownershipGained); // we leave if we gain ownership of readPosition
+
+    index = value.getIndex();
+    return true;
+}
+
+template <uint64_t Capacity, typename ValueType>
+bool IndexQueue<Capacity, ValueType>::popIfFull(ValueType& index)
+{
+    // we do NOT need a CAS loop here since if we detect that the queue is not full
+    // someone else popped an element and we do not retry to check whether it was filled AGAIN
+    // concurrently (which will usually not be the case and then we would return false anyway)
+    // if it is filled again we can (and will) retry popIfFull from the call site
+
+    // the queue is full if and only if write position and read position are the same but read position is
+    // one cycle behind write position
+    // unfortunately it seems impossible in this design to check this condition without loading
+    // write posiion and read position (which causes more contention)
+
+    auto writePosition = loadNextWritePosition();
+    auto readPosition = loadNextReadPosition();
+    auto value = loadValueAt(readPosition);
+
+    auto isFull = writePosition.getIndex() == readPosition.getIndex() && readPosition.isOneCycleBehind(writePosition);
+
+    if (isFull)
+    {
+        if (tryToGainOwnershipAt(readPosition))
+        {
+            index = value.getIndex();
+            return true;
+        }
+    }
+
+    // otherwise someone else has dequeued an index and the queue was not full at the start of this popIfFull
+    return false;
+}
+
+template <uint64_t Capacity, typename ValueType>
+bool IndexQueue<Capacity, ValueType>::empty()
+{
+    auto oldReadIndex = m_readPosition.load(std::memory_order_relaxed);
+    auto value = m_cells[oldReadIndex.getIndex()].load(std::memory_order_relaxed);
+
+    // if m_readPosition is ahead by one cycle compared to the value stored at head,
+    // the queue was empty at the time of the loads above (but might not be anymore!)
+    return value.isOneCycleBehind(oldReadIndex);
+}
+
+template <uint64_t Capacity, typename ValueType>
+typename IndexQueue<Capacity, ValueType>::Index IndexQueue<Capacity, ValueType>::loadNextReadPosition() const
```

Comment:
Please get rid of all these unnecessary atomic abstractions. This makes the code ultra hard to review since you have to jump around all the time and brings no benefit. Lock free code is hard to understand but when you call
```
m_readPosition.load()
```
it is pretty clear to me that you loadNextReadPosition. Its the same which is already written above. But the memory order is important. I would like to understand and confirm that your memory_order_relaxed is correct but this jumping around makes it hard for me to understand

Reply:
A matter that can be argued back and forth, unfortunately. It depends on the abstraction level we have in mind when we look at the code. I will take a look whether this can be made easier, but we may need to discuss this. @elfenpiff